docs/src/main/asciidoc/se/injection.adoc

```diff
@@ -0,0 +1,677 @@
+
+[source,java]
+.Asynchronous Event Producer Example
+----
+include::{sourcedir}/se/inject/EventsExample.java[tag=snippet_4, indent=0]
+----
+
+=== Asynchronous Observer
+Asynchronous observer methods are invoked from separate threads (through the executor service mentioned above), and their results are ignored by the Event Emitter; if there is an exception thrown from the observer method, it is logged with `WARNING` log level into logger named `io.helidon.service.registry.EventManager`.
+
+To declare an asynchronous observer use annotation `Event.AsyncObserver` instead of `Event.Observer`.
+
+[source,java]
+.Asynchronous Event Observer Example
+----
+include::{sourcedir}/se/inject/EventsExample.java[tag=snippet_5, indent=0]
+----
+
+== Programmatic Lookup
+
+There are two primary ways to access services from the service registry:
+
+- *Dependency Injection* – Services are injected automatically at defined injection points.
+- *Programmatic Lookup* – Services are retrieved manually by accessing the service registry directly.
+
+If you want to use programmatic lookup, there are two main ways to access a `ServiceRegistry` instance:
+
+- Create a new ServiceRegistry instance and search for the desired service.
+- Inject the existing ServiceRegistry instance currently used by Helidon and retrieve services from it.
+
+To create a registry instance:
+
+[source,java]
+----
+include::{sourcedir}/se/inject/ServiceRegistryExample.java[tag=snippet_1, indent=0]
+----
+
+Note that all instances are created lazily, meaning the service registry does nothing by default.
+If a service performs any actions during construction or post-construction,
+you must first retrieve an instance from the registry to trigger its initialization.
+
+Special registry operations:
+
+- `List<ServiceInfo> lookupServices(Lookup lookup)` - get all service descriptors that match the lookup
+- `Optional<T> get(ServiceInfo)` - get an instance for the provided service descriptor
+
+The common registry operations are grouped by method name. Acceptable parameters are described below.
+
+Registry methods:
+
+- `T get(...)` - immediately get an instance of a contract from the registry; throws if implementation not available
+- `Optional<T> first(...)` - immediately get an instance of a contract from the registry; there may not be an implementation
+available
+- `List<T> all(...)` - immediately get all instances of a contract from the registry; result may be empty
+- `Supplier<T> supply(...)` - get a supplier of an instance; the service may be instantiated only when `get` is called
+- `Supplier<Optional<T>> supplyFirst(...)` - get a supplier of an optional instance
+- `Supplier<List<T>> supplyAll(...)` - get a supplier of all instances
+
+Lookup parameter options:
+
+- `Class<?>` - the contract we are looking for
+- `TypeName` - the same, but using Helidon abstraction of type names (may have type arguments)
+- `Lookup` - a full search criteria for a registry lookup
+
+== Startup
```

Comment:
This seems to be a bit unfinished, as we need explanation of `RunLevel` and how it is used in generated main class, and in `io.helidon.Main`.